File: tests/test_leduk.py
import unittest

from oz.game.leduk import LedukPoker, Action, Card


class TestLedukPoker(unittest.TestCase):

    def test_init(self):
        g = LedukPoker()

    def test_deal(self):
        g = LedukPoker()
        self.assertIsNotNone(g.hand[0])
        self.assertIsNotNone(g.hand[1])
        self.assertEqual(len(g.deck), 4)
        self.assertIsNone(g.board)
        self.assertEqual(g.round, 0)

    def test_bet(self):
        g = LedukPoker()
        g.act(Action.Bet)
        g.act(Action.Call)
        self.assertEqual(g.pot[0], 3)
        self.assertEqual(g.pot[1], 3)
        self.assertEqual(g.round, 1)

    def test_check(self):
        g = LedukPoker()
        g.act(Action.Check)
        g.act(Action.Call)
        self.assertEqual(g.pot[0], 1)
        self.assertEqual(g.pot[1], 1)
        self.assertEqual(g.round, 1)
        self.assertIsNotNone(g.board)

    def test_fold(self):
        g = LedukPoker()
        g.act(Action.Bet)
        g.act(Action.Fold)
        self.assertEqual(g.pot[0], 3)
        self.assertEqual(g.pot[1], 1)
        self.assertTrue(g.is_terminal())
        self.assertEqual(g.reward(), 1)

    def test_reraise(self):
        g = LedukPoker()
        g.act(Action.Bet)
        g.act(Action.Bet)
        with self.assertRaises(ValueError) as ex:
            g.act(Action.Bet)

    def test_bet_rounds(self):
        g = LedukPoker()
        g.act(Action.Bet)
        g.act(Action.Raise)
        g.act(Action.Call)
        g.act(Action.Bet)
        g.act(Action.Raise)
        g.act(Action.Call)
        self.assertEqual(g.pot[0], 13)
        self.assertEqual(g.pot[1], 13)
        self.assertTrue(g.is_terminal())

    def test_hand_rank(self):
        g = LedukPoker()
        self.assertGreater(
            g._rank_hand(card=Card.Jack, board=Card.Jack),
            g._rank_hand(card=Card.King, board=Card.Queen))
        self.assertGreater(
            g._rank_hand(card=Card.King,  board=Card.Queen),
            g._rank_hand(card=Card.Queen, board=Card.King))

    def test_reward(self):
        # P1 Win
        g = LedukPoker()
        g.act(Action.Bet)
        g.act(Action.Call)
        g.act(Action.Bet)
        g.act(Action.Call)
        g.hand[0] = Card.King
        g.hand[1] = Card.Queen
        g.board = Card.Jack
        self.assertEqual(g.reward(), 7)

        # P2 Win
        g = LedukPoker()
        g.act(Action.Bet)
        g.act(Action.Raise)
        g.act(Action.Call)
        g.act(Action.Bet)
        g.act(Action.Raise)
        g.act(Action.Call)
        g.hand[0] = Card.King
        g.hand[1] = Card.Queen
        g.board = Card.Queen
        self.assertEqual(g.reward(), -13)

        # Draw
        g = LedukPoker()
        g.act(Action.Bet)
        g.act(Action.Call)
        g.act(Action.Bet)
        g.act(Action.Call)
        g.hand[0] = Card.Queen
        g.hand[1] = Card.Queen
        g.board = Card.King
        self.assertEqual(g.reward(), 0)

    def test_legal_actions(self):
        g = LedukPoker()
        g.act(Action.Call)
        actions = g.legal_actions()
        self.assertIn(Action.Call, actions)
        self.assertIn(Action.Fold, actions)
        self.assertIn(Action.Raise, actions)

        g = LedukPoker()
        g.act(Action.Call)
        g.act(Action.Raise)
        g.act(Action.Raise)
        actions = g.legal_actions()
        self.assertNotIn(Action.Raise, actions)

File: oz/csrc/tests/test_leduk.cpp
#include <catch.hpp>

#include "game.h"
#include "games/leduk.h"

using oz::leduk_poker_t;
using oz::CHANCE;
using oz::P1;
using oz::P2;

using card_t = leduk_poker_t::card_t;
using action_t = leduk_poker_t::action_t;

TEST_CASE("leduk poker basic actions", "[leduk]") {
  leduk_poker_t g;

  CHECK(g.player() == CHANCE);
  g.act_(action_t::J1);
  CHECK(g.player() == CHANCE);
  g.act_(action_t::Q2);
  CHECK(g.player() == P1);
  g.act_(action_t::Raise);
  CHECK(g.player() == P2);
  g.act_(action_t::Call);
  CHECK(g.player() == CHANCE);
}

leduk_poker_t _leduk_after_deal() {
  auto g = leduk_poker_t();
  g.act_(action_t::J1);
  g.act_(action_t::Q2);
  return g;
}

TEST_CASE("leduk poker bet", "[leduk]") {
  leduk_poker_t g = _leduk_after_deal();

  g.act_(action_t::Raise);
  g.act_(action_t::Call);

  CHECK(g.pot(P1) == 3);
  CHECK(g.pot(P2) == 3);
  CHECK(g.round() == 1);
}

TEST_CASE("leduk poker check", "[leduk]") {
  leduk_poker_t g = _leduk_after_deal();

  g.act_(action_t::Call);
  g.act_(action_t::Call);

  CHECK(g.pot(P1) == 1);
  CHECK(g.pot(P2) == 1);
  CHECK(g.round() == 1);
  CHECK(g.board() == card_t::NA);
}

TEST_CASE("leduk poker fold", "[leduk]") {
  leduk_poker_t g = _leduk_after_deal();

  g.act_(action_t::Raise);
  g.act_(action_t::Fold);

  CHECK(g.pot(P1) == 3);
  CHECK(g.pot(P2) == 1);
  CHECK(g.is_terminal());
  CHECK(g.utility(P1) == 1);
}

TEST_CASE("leduk poker reraise", "[leduk]") {
  leduk_poker_t g = _leduk_after_deal();

  g.act_(action_t::Raise);
  g.act_(action_t::Raise);

  // TODO
}

TEST_CASE("leduk poker bet rounds", "[leduk]") {
  leduk_poker_t g = _leduk_after_deal();

  g.act_(action_t::Raise);
  g.act_(action_t::Raise);
  g.act_(action_t::Call);
  CHECK(g.player() == CHANCE);
  g.act_(action_t::K);
  g.act_(action_t::Raise);
  g.act_(action_t::Raise);
  g.act_(action_t::Call);

  CHECK(g.pot(P1) == 13);
  CHECK(g.pot(P2) == 13);
  CHECK(g.is_terminal());
}

TEST_CASE("leduk poker bet reward", "[leduk]") {
  leduk_poker_t g = leduk_poker_t();

  g.act_(action_t::K1);
  g.act_(action_t::Q2);
  g.act_(action_t::Raise);
  g.act_(action_t::Call);
  g.act_(action_t::J);
  g.act_(action_t::Raise);
  g.act_(action_t::Call);

  CHECK(g.hand(P1) == card_t::King);
  CHECK(g.hand(P2) == card_t::Queen);
  CHECK(g.board() == card_t::Jack);
  CHECK(g.is_terminal());
  CHECK(g.utility(P1) == 7);

  g = leduk_poker_t();

  g.act_(action_t::K1);
  g.act_(action_t::Q2);
  g.act_(action_t::Raise);
  g.act_(action_t::Raise);
  g.act_(action_t::Call);
  g.act_(action_t::Q);
  g.act_(action_t::Raise);
  g.act_(action_t::Raise);
  g.act_(action_t::Call);

  CHECK(g.hand(P1) == card_t::King);
  CHECK(g.hand(P2) == card_t::Queen);
  CHECK(g.board() == card_t::Queen);
  CHECK(g.is_terminal());
  CHECK(g.utility(P1) == -13);

  g = leduk_poker_t();

  g.act_(action_t::Q1);
  g.act_(action_t::Q2);
  g.act_(action_t::Raise);
  g.act_(action_t::Call);
  g.act_(action_t::K);
  g.act_(action_t::Raise);
  g.act_(action_t::Call);

  CHECK(g.hand(P1) == card_t::Queen);
  CHECK(g.hand(P2) == card_t::Queen);
  CHECK(g.board() == card_t::King);
  CHECK(g.is_terminal());
  CHECK(g.utility(P1) == 0);
}
